### PR TITLE
to make repetition penalty faster

### PR DESCRIPTION
**Pull request by ccrhx4**
This PR is to fix very slow sampling process when repetition penalty is set.

The fix includes:
1. Enable pin_memory on HPU
2. Padding prompt tokens and output_tokens to avoid recompile
3. Replace slow ops

Before the fix:
SamplingParams(n=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.06, temperature=1.0, top_p=1.0, top_k=-1, min_p=0.0, seed=None, stop=[], stop_token_ids=[], include_stop_str_in_output=False, ignore_eos=True, max_tokens=1024, min_tokens=0, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None), guided_decoding=None
Warming up...
Profiling iterations: 100%|5/5 [03:24<00:00, 40.99s/it]
Avg latency: 40.98862759781768 seconds
10% percentile latency: 11.699748958216514 seconds
25% percentile latency: 11.73845003999304 seconds
50% percentile latency: 11.801458386995364 seconds
75% percentile latency: 11.861465670051984 seconds
90% percentile latency: 99.46527566103033 seconds
99% percentile latency: 152.02756165561732 seconds

After the fix:
SamplingParams(n=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.06, temperature=1.0, top_p=1.0, top_k=-1, min_p=0.0, seed=None, stop=[], stop_token_ids=[], include_stop_str_in_output=False, ignore_eos=True, max_tokens=1024, min_tokens=0, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None), guided_decoding=None
Warming up...
Profiling iterations: 100%| 5/5 [00:57<00:00, 11.59s/it]
Avg latency: 11.58703240059549 seconds
10% percentile latency: 11.444069900200702 seconds
25% percentile latency: 11.511425047006924 seconds
50% percentile latency: 11.525146245025098 seconds
75% percentile latency: 11.556680046953261 seconds
90% percentile latency: 11.788318535778672 seconds
99% percentile latency: 11.927301629073918 seconds

Testing code is by: https://github.com/ccrhx4/huanxing.vllm-fork/blob/slow_repetition_penalty/benchmarks/reproduce.sh

### CODE REVIEW

File: vllm/worker/cache_engine.py [hunk 1]
```diff
@@ -78,11 +79,18 @@ def _allocate_kv_cache(
             # null block in CpuGpuBlockAllocator requires at least that
             # block to be zeroed-out.
             # We zero-out everything for simplicity.
-            kv_cache.append(
-                torch.zeros(kv_cache_shape,
-                            dtype=self.dtype,
-                            pin_memory=pin_memory,
-                            device=device))
+            if pin_memory:
```
Comment (michalkuligowski) [inline vllm/worker/cache_engine.py:82]: Why is this if needed?

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: vllm/worker/cache_engine.py [hunk 2]
```diff
@@ -78,11 +79,18 @@ def _allocate_kv_cache(
...
+                    kv_cache.append(
+                        torch.zeros(kv_cache_shape,
+                                dtype=self.dtype,
+                                pin_memory=pin_memory,
+                                device=device))
+                else:
+                    kv_cache.append(
+                        torch.zeros(kv_cache_shape,
+                                dtype=dtype,
+                                device=device).pin_memory(device="hpu"))
```
Comment (michalkuligowski) [inline vllm/worker/cache_engine.py:93]: This breaks non hpu devices?

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: vllm/worker/cache_engine.py (hunk 1 above)
Comment (ccrhx4) [reply vllm/worker/cache_engine.py]: I have reverted all the changes in this file, as this _allocate_kv_cache() was not used in HPU.

File: vllm/worker/cache_engine.py (hunk 2 above)
Comment (ccrhx4) [reply vllm/worker/cache_engine.py]: see above

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: vllm/model_executor/sampling_metadata.py [hunk 3]
```diff
@@ -508,19 +510,21 @@ def from_lists(
         do_penalties = prompt_tokens or output_tokens
 
         if do_penalties:
-            prompt_t = make_tensor_with_pad(
+            prompt_t = make_tensor_with_pad_align(
```
[the remark this answers is not in the document]
Comment (michalkuligowski) [reply vllm/model_executor/sampling_metadata.py:513]: Also please fix formatting issues

[the remark this answers is not in the document]
Comment (ccrhx4) [reply vllm/model_executor/sampling_metadata.py:513]: Hi Michal, 

I need to change the parameters of make_tensor_with_pad to pass a new input. But the origin make_tensor_with_pad is used in many places and that may interfere with other logic, and make the code hard to maintain.

How about I change like this:

```
if do_penalties:
            if is_hpu():
                prompt_t = make_tensor_with_pad_align(
                    prompt_tokens,
                    vocab_size,
                    device="cpu",
                    dtype=torch.int64,
                    pin_memory=pin_memory,
                    max_len_align=1024,
                )
                output_t = make_tensor_with_pad_align(
                    output_tokens,
                    vocab_size,
                    device="cpu",
                    dtype=torch.int64,
                    pin_memory=pin_memory,
                    max_len_align=1024,
                )
            else:
                prompt_t = make_tensor_with_pad(
                    prompt_tokens,
                    vocab_size,
                    device="cpu",
                    dtype=torch.int64,
                    pin_memory=pin_memory,
                )
                output_t = make_tensor_with_pad(
                    output_tokens,
                    vocab_size,
                    device="cpu",
                    dtype=torch.int64,
                    pin_memory=pin_memory,
                )
```

File: vllm/utils.py [hunk 4]
```diff
@@ -833,10 +854,40 @@ def make_tensor_with_pad(
...
+
+    The padding is applied to the end of each inner list until it reaches
+    `max_len`.
+    """
+    np_dtype = TORCH_DTYPE_TO_NUMPY_DTYPE[dtype]
+    padded_x = make_ndarray_with_pad_align(x, pad, np_dtype, 
+                max_len_align=max_len_align)
+
+    tensor = torch.from_numpy(padded_x).to(device)
+    if pin_memory:
```
Comment (michalkuligowski) [inline vllm/utils.py:884]: if not needed, since this method is called for hpu

File: vllm/utils.py [hunk 5]
```diff
@@ -833,10 +854,40 @@ def make_tensor_with_pad(
 
     tensor = torch.from_numpy(padded_x).to(device)
     if pin_memory:
-        tensor = tensor.pin_memory()
+        if not current_platform.is_hpu():
+            tensor = tensor.pin_memory()
+        else:
+            tensor = tensor.pin_memory("hpu")
```
Comment (michalkuligowski) [inline vllm/utils.py:860]: can be removed, as it wont be called now

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: vllm/utils.py (hunk 5 above)
Comment (ccrhx4) [reply vllm/utils.py:860]: Hi Michal, this method make_tensor_with_pad is still called from different places. It is replaced by make_tensor_with_pad_align in the repetition penaly. So I think we still need the check here.

File: vllm/utils.py (hunk 4 above)
Comment (ccrhx4) [reply vllm/utils.py:884]: HI Michal, I removed the device check logic here and kept the pin_memory check. In this way, this method behavior is exactly the same to the un-aligned version.

File: vllm/utils.py [hunk 6]
```diff
@@ -812,6 +810,29 @@ def make_ndarray_with_pad(
...
+) -> npt.NDArray:
+    """
+    Make a padded array from 2D inputs.
+
+    The padding is applied to the end of each inner list until it reaches
+    `max_len`.
+    """
+    # Unlike for most functions, map is faster than a genexpr over `len`
+    max_len = max(map(len, x), default=0)
+    max_len_aligned =  math.ceil(max_len / max_len_align) * max_len_align
```
Comment (kdamaszk) [inline vllm/utils.py:828]: There are some mypy failures, please fix it

File: vllm/model_executor/sampling_metadata.py [hunk 7]
```diff
@@ -522,20 +524,38 @@ def from_lists(
...
-                pin_memory=pin_memory,
-            )
+            if current_platform.is_hpu():
+                prompt_t = make_tensor_with_pad_align(
+                    prompt_tokens,
+                    vocab_size,
+                    device="cpu",
+                    dtype=torch.int64,
+                    pin_memory=pin_memory,
+                    max_len_align=1024,
```
Comment (kdamaszk) [inline vllm/model_executor/sampling_metadata.py:534]: why max_len_align is set to 1024?

File: vllm/utils.py (hunk 6 above)
Comment (ccrhx4) [reply vllm/utils.py]: Fixed the latest commit.

File: vllm/model_executor/sampling_metadata.py (hunk 7 above)
Comment (ccrhx4) [reply vllm/model_executor/sampling_metadata.py:533]: it will padding the penalty tensor to the max_len in the array, and max_len is rounded to the nearest 1024. It is to avoid recompile and also not to bring too much computation because of the padding.

**Comment by kdamaszk** [comment]
@ccrhx4 please check CI status -- there are still some formatting issues with yapf which should be easy to fix.
However, there is also more important issue with CPU tests -- please note that they are using fake_hpu (HPU path simulated on CPU). You need to cover it properly -- you can use `from vllm.utils import is_fake_hpu`.
To reproduce on local machine please run:
```
VLLM_SKIP_WARMUP=true VLLM_PROMPT_SEQ_BUCKET_MAX=128 VLLM_USE_FAKE_HPU=1 python examples/offline_inference_fakehpu.py
```

**Comment by ccrhx4** [comment]
> @ccrhx4 please check CI status -- there are still some formatting issues with yapf which should be easy to fix. However, there is also more important issue with CPU tests -- please note that they are using fake_hpu (HPU path simulated on CPU). You need to cover it properly -- you can use `from vllm.utils import is_fake_hpu`. To reproduce on local machine please run:
> 
> ```
> VLLM_SKIP_WARMUP=true VLLM_PROMPT_SEQ_BUCKET_MAX=128 VLLM_USE_FAKE_HPU=1 python examples/offline_inference_fakehpu.py
> ```

I have updated the code to fix the cpu testing and code format. Please help to trigger the test. Thanks!

**Comment by ccrhx4** [comment]
Jenkins UI failed, but there were no error log. Please help. @kdamaszk Thanks!

**Comment by kdamaszk** [comment]
@ccrhx4 please rebase your review and we will rerun CI

**Review by kdamaszk — APPROVED** [review]

**Comment by kdamaszk** [comment]
@ccrhx4 your change introduced regression in LoRA tests, we need to revert it. Please revalidate and fix your change including LoRA specifics.

**Comment by ccrhx4** [comment]
@kdamaszk I see. I will update my code.

**Comment by ccrhx4** [comment]
@kdamaszk I will submit a new PR to include only changes in sampling logic and do not touch pin_memory. pin_memory support on HPU is not mature enough, it will make the code dirty if I follow HPU's way (only HPU goes this way.)

I am submiting JIRAs asking for pin_memory support in default pytorch way on HPU.
